amazon-kinesis-client/src/main/java/software/amazon/kinesis/coordinator/Scheduler.java

```diff
@@ -345,6 +357,31 @@ void initialize() {
         }
     }
 
+    @VisibleForTesting
+    boolean shouldInitiateLeaseSync() throws InterruptedException,
+            DependencyException, ProvisionedThroughputException, InvalidStateException {
+        long waitTime = ThreadLocalRandom.current().nextLong(MIN_WAIT_TIME_FOR_LEASE_TABLE_CHECK_MILLIS, MAX_WAIT_TIME_FOR_LEASE_TABLE_CHECK_MILLIS);
+        long waitUntil = System.currentTimeMillis() + waitTime;
+
+        boolean isLeaseTableEmpty = true;
+        while (System.currentTimeMillis() < waitUntil && (isLeaseTableEmpty = leaseRefresher.isLeaseTableEmpty())) {
+            // check every 3 seconds if lease table is still empty,
+            // to minimize contention between all workers bootstrapping at the same time
+            log.info("Lease table is still empty. Checking again in {} ms", LEASE_TABLE_CHECK_FREQUENCY_MILLIS);
+            Thread.sleep(LEASE_TABLE_CHECK_FREQUENCY_MILLIS);
+        }
+        return isLeaseTableEmpty;
+    }
+
+    private void waitUntilHashRangeCovered() throws InterruptedException {
```

Comment:
Should this be taking streamName as part of this to support multistream. If this is not handled can we remove this code ?

Reply:
leaderElectedPeriodicShardSyncManager has the reference of the streamToShardSyncTaskManagerMap, which contains the streamIdentifier of all streams. In the leaderElectedPeriodicShardSyncManager.hashRangeCovered implementation, we would go over all streams and check if hash range is covered.

Reply:
streamToShardSyncTaskManagerMap might not reflect the up to date snapshot of the active streams. We should always check the currentStreamConfigMap to determine the streams to be checked. Also they should be executed with lock on  currentStreamConfigMap. Let's add a TODO.